### PR TITLE
Type check arguments and return types (BT-671)

### PR DESCRIPTION
**Pull request by jamesc**
## Summary

Implements ADR 0025 Phase 2 argument type checking at call sites and return type checking in method bodies.

**Linear issue:** https://linear.app/beamtalk/issue/BT-671

## Key Changes

- **`check_argument_types`**: Validates argument types against `MethodInfo.param_types` from `ClassHierarchy`. Dynamic arguments never produce warnings.
- **`check_return_type`**: Compares inferred method body return type against declared `return_type`. Skips `@primitive` methods.
- **`is_type_compatible`**: Subtype checking via superclass chain (e.g., Integer valid for Number).
- **`check_override_param_compatibility`**: Warns when child method param types are incompatible with parent's.
- **Duplicate warning prevention**: Binary messages skip `check_argument_types` since `check_binary_operand_types` already provides specific warnings.
- **`asType:`** cast continues to suppress subsequent type warnings (already worked).
- All diagnostics are `Severity::Warning` (never errors, per ADR 0025).

## Tests

17 new tests covering all acceptance criteria:
- `'hello' + 42` warns; `3 + 4` does not
- `getBalance -> Integer => 'oops'` warns
- Integer valid for Number (subtype)
- Dynamic args never warn
- Override incompatible param types warn
- Empty body / primitive methods handled correctly
- Exactly 1 warning for binary ops (no duplicates)
- Return type subtype compatibility

## Files Modified

- `crates/beamtalk-core/src/semantic_analysis/type_checker.rs`

## Summary by CodeRabbit

* **New Features**
  * Enhanced type validation: argument-type checks for message sends, return-type validation for method bodies, and override-compatibility checks across class hierarchies; emits user-facing warnings with actionable hints when mismatches occur.
* **Tests**
  * Added extensive tests for type compatibility, binary operand handling, dynamic/symbolic types, primitive skips, override compatibility, and related edge cases to improve reliability.

### CODE REVIEW

File: crates/beamtalk-core/src/semantic_analysis/type_checker.rs [hunk 1]
```diff
@@ -146,9 +149,11 @@ impl TypeChecker {
                 method_env.in_class_method = true;
                 method_env.set("self", InferredType::Known(class.name.name.clone()));
                 Self::set_param_types(&mut method_env, &method.parameters);
+                let mut body_type = InferredType::Dynamic;
                 for expr in &method.body {
-                    self.infer_expr(expr, hierarchy, &mut method_env, is_abstract);
+                    body_type = self.infer_expr(expr, hierarchy, &mut method_env, is_abstract);
                 }
+                self.check_return_type(method, &body_type, &class.name.name, hierarchy);
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-core/src/semantic_analysis/type_checker.rs]: Fixed in same commit — all three body loops now break on early return.
